homeassistant/components/smartthings/smartapp.py

```diff
@@ -232,44 +232,52 @@ def setup_smartapp(hass, app):
-    to receive the related attribute updates.  Finally, create a config entry
-    representing the installation if this is not the first installation under
-    the account.
+    Create a config entry representing the installation if this is not
+    the first installation under the account, otherwise store the data
+    for the config flow.
     """
-    await smartapp_sync_subscriptions(
-        hass, req.auth_token, req.location_id, req.installed_app_id,
-        skip_delete=True)
-
-    # The permanent access token is copied from another config flow with the
-    # same parent app_id.  If one is not found, that means the user is within
-    # the initial config flow and the entry at the conclusion.
-    access_token = next((
-        entry.data.get(CONF_ACCESS_TOKEN) for entry
+    install_data = {
+        CONF_INSTALLED_APP_ID: req.installed_app_id,
+        CONF_LOCATION_ID: req.location_id,
+        CONF_REFRESH_TOKEN: req.refresh_token
+    }
+    # App attributes (client id/secret, etc...) are copied from another entry
+    # with the same parent app_id.  If one is not found, the install data is
+    # stored for the config flow to retrieve during the wait step.
+    entry = next((
+        entry for entry
         in hass.config_entries.async_entries(DOMAIN)
         if entry.data[CONF_APP_ID] == app.app_id), None)
-    if access_token:
+    if entry:
+        data = entry.data.copy()
+        data.update(install_data)
         # Add as job not needed because the current coroutine was invoked
         # from the dispatcher and is not being awaited.
         await hass.config_entries.flow.async_init(
             DOMAIN, context={'source': 'install'},
-            data={
-                CONF_APP_ID: app.app_id,
-                CONF_INSTALLED_APP_ID: req.installed_app_id,
-                CONF_LOCATION_ID: req.location_id,
-                CONF_ACCESS_TOKEN: access_token
-            })
+            data=data)
+    else:
+        # Store the data where the flow can find it
+        hass.data[DOMAIN][CONF_INSTALLED_APPS].append(install_data)
 
 
 async def smartapp_update(hass: HomeAssistantType, req, resp, app):
     """
     Handle when a SmartApp is updated (reconfigured) by the user.
 
-    Synchronize subscriptions to ensure we're up-to-date.
+    Store the refresh token and reload the config entry.
     """
-    await smartapp_sync_subscriptions(
-        hass, req.auth_token, req.location_id, req.installed_app_id)
+    # Update refresh token in config entry
+    entry = next((entry for entry in hass.config_entries.async_entries(DOMAIN)
+                  if entry.data.get(CONF_INSTALLED_APP_ID) ==
+                  req.installed_app_id),
+                 None)
+    if entry:
+        entry.data[CONF_REFRESH_TOKEN] = req.refresh_token
+        hass.config_entries.async_update_entry(entry)
+        # Reload it
+        await entry.async_unload(hass)
```

Comment:
Do we need to do this? Are we allowed to interact directly with the entry like this?

Reply:
It provides a way for the user to update the devices/entities in HASS when changed in SmartThings without having to restart HASS.  I also plan to expose this as a service in the future so it can be done ad-hoc.  I'll improve the comment on why it's being reloaded.  As far as whether we can do this, I don't see why not as along as all the platforms support unloading.  Config entries are meant to be loaded/unloaded at runtime.

Reply:
Sorry, I'm asking about the api, not the result or necessity of the result.

Since we have `hass.config_entries.async_update_entry` for updating entro, I would expect us to have corresponding methods for other changes to an entry. If we don't have those, maybe we should have a discussion somewhere about that.

Reply:
Good question -- `config_entries` could add `async_unload` and `async_reload` to be more complete.  @balloob What do you think?  Fine for now or should we add it to core?

Reply:
Martin is correct, we should not directly interact with the entry, as that might not invoke the right listeners. We can add an `unload`, as we already have a `load`.

Reply:
Ok I’ll open an arch issue to dicuss and remove  entity reloading from this PR so it can move forward.  It’s not critical.